src/shared/lib/AnnotationUtils.ts

```diff
@@ -45,18 +50,129 @@ export function geneToMyCancerGenome(myCancerGenomes:IMyCancerGenome[]):IMyCance
     return map;
 }
 
-export function geneAndProteinPosToHotspots(hotspots:HotspotMutation[]):{[key:string]: boolean}
+export function geneAndProteinPosToHotspots(hotspots:HotspotMutation[]):{[key:string]: HotspotMutation}
 {
     // key: geneSymbol_proteinPosition
     // (protienPosition: start[_end])
-    const map: {[key:string]: boolean} = {};
+    const map: {[key:string]: HotspotMutation} = {};
 
     // create a map for a faster lookup
     hotspots.forEach(function(hotspot:HotspotMutation) {
         const positions = hotspot.residue.match(/[0-9]+/g) || []; // start (and optionally end) positions
         const key = [hotspot.hugoSymbol.toUpperCase()].concat(positions).join("_");
-        map[key] = true;
+        map[key] = hotspot;
     });
 
     return map;
 }
+
+export function indexHotspots(hotspots:HotspotMutation[], defaultHotspotType?: string): IHotspotIndex
+{
+    const index:IHotspotIndex = {};
+
+    hotspots.forEach((hotspot:HotspotMutation) => {
+        let geneIndex = index[hotspot.hugoSymbol.toUpperCase()];
+
+        if (!geneIndex)
+        {
+            geneIndex = {};
+
+            index[hotspot.hugoSymbol.toUpperCase()] = geneIndex;
+        }
+
+        const hotspotType = hotspot.type || defaultHotspotType || "hotspot";
+        let typeIndex = index[hotspot.hugoSymbol.toUpperCase()][hotspotType];
+
+        if (!typeIndex)
+        {
+            typeIndex = {
+                hotspotMutations: [],
+                hotspotSet: new HotspotSet([]) // this is a placeholder to avoid TS errors.
+                                               // overwritten by the initHotspotSets function.
+            };
+
+            index[hotspot.hugoSymbol.toUpperCase()][hotspotType] = typeIndex;
+        }
+
+        typeIndex.hotspotMutations.push(hotspot);
+    });
+
+    initHotspotSets(index);
+
+    return index;
+}
+
+export function initHotspotSets(hotspotIndex: IHotspotIndex)
+{
+    _.keys(hotspotIndex).forEach((geneSymbol:string) => {
+        _.keys(hotspotIndex[geneSymbol]).forEach((hotspotType:string) => {
+            const intervals = hotspotIndex[geneSymbol][hotspotType].hotspotMutations.map(
+                (hotspot:HotspotMutation):[number, number] => {
+                    // start (and optionally end) positions
+                    const positions = hotspot.residue.match(/[0-9]+/g) || [];
```

Comment:
instead of parsing the residue, we should use `hotspot.aminoAcidPosition.start` and `hotspot.aminoAcidPosition.end` values